File: db4oj/core/src/com/db4o/inside/btree/BTreeRangeIterator.java
/**
 * 
 */
package com.db4o.inside.btree;

import com.db4o.Transaction;
import com.db4o.foundation.*;

class BTreeRangeIterator implements Iterator4 {
	
	private final BTreeRangeSingle _range;
	
	private BTreePointer _cursor;
	private BTreeNode _lastNode = null;
	private BTreePointer _current;
	
	public BTreeRangeIterator(BTreeRangeSingle range) {
		_range = range;
		_cursor = range.first();
	}

	public boolean moveNext() {
		
		while(! reachedEnd(_cursor)){
            
            BTreeNode node = _cursor.node();
            
            if(node != _lastNode){
                node.prepareWrite(transaction());
                
                // Alternative: work in read mode, hold the reader here.
                
                _lastNode = node;
            }
            
            Object obj = _cursor.key();
            
            if(obj != No4.INSTANCE){
                _current = _cursor;
                _cursor = _cursor.next();
                return true;
            }
            
            _cursor = _cursor.next();
        }
		_current = null;
		return false;
	}
	
	public Object current() {
		return currentPointer().key();
	}

	private BTreePointer currentPointer() {
		if (null == _current) {
			throw new IllegalStateException();
		}
		return _current;
	}
	
	private boolean reachedEnd(BTreePointer cursor){
        if(cursor == null){
            return true;
        }
        if(_range.end() == null){
            return false;
        }
        return _range.end().equals(cursor);
    }
    
    private Transaction transaction(){
        return _range.transaction();
    }
}

File: db4on/core/src/core/com/db4o/inside/btree/BTreeRangeIterator.cs
namespace com.db4o.inside.btree
{
	internal class BTreeRangeIterator : com.db4o.foundation.KeyValueIterator
	{
		private readonly com.db4o.inside.btree.BTreeRangeImpl _range;

		private com.db4o.inside.btree.BTreePointer _cursor;

		private com.db4o.inside.btree.BTreeNode _lastNode = null;

		private com.db4o.inside.btree.BTreePointer _current;

		public BTreeRangeIterator(com.db4o.inside.btree.BTreeRangeImpl range)
		{
			_range = range;
			_cursor = range.First();
		}

		public virtual bool MoveNext()
		{
			while (!ReachedEnd(_cursor))
			{
				com.db4o.inside.btree.BTreeNode node = _cursor.Node();
				if (node != _lastNode)
				{
					node.PrepareWrite(Transaction());
					_lastNode = node;
				}
				object obj = _cursor.Key();
				if (obj != com.db4o.foundation.No4.INSTANCE)
				{
					_current = _cursor;
					_cursor = _cursor.Next();
					return true;
				}
				_cursor = _cursor.Next();
			}
			_current = null;
			return false;
		}

		public virtual object Key()
		{
			return Current().Key();
		}

		public virtual object Value()
		{
			return Current().Value();
		}

		private com.db4o.inside.btree.BTreePointer Current()
		{
			if (null == _current)
			{
				throw new System.InvalidOperationException();
			}
			return _current;
		}

		private bool ReachedEnd(com.db4o.inside.btree.BTreePointer cursor)
		{
			if (cursor == null)
			{
				return true;
			}
			if (_range.End() == null)
			{
				return false;
			}
			return _range.End().Equals(cursor);
		}

		private com.db4o.Transaction Transaction()
		{
			return _range.Transaction();
		}
	}
}
